### DISCUSSION TITLE
Cannot enable CORS for my API in vercel Serverless functions: Response not sending in headers. Works in local environment

### DISCUSSION THREAD

**Discussion by dBianchii** [vercel/community#3274] [Help] [answered]
### Summary

I am trying to call my api from another website.
I have one Next.JS Serverless function in _/api/trpc/_  and I am trying to call it from another deployed frontend application also hosted in vercel. I have provided the link to a client application I am using to call my api. Take a look inside https://client-nextjs-one.vercel.app/client and access network tab on developer tools. You will see that I am trying to make a request, but am being blocked by a CORS issue.

### Example

https://client-nextjs-one.vercel.app/client

### Steps to Reproduce

Hello, I have come across many people on the internet with this same issue. I am creating an endpoint that should allow CORS. But my response is not sending the headers I want.

I am trying to call my TRPC api from another website, thus needing to enable Cross-Origin-Resource-Sharing. (CORS)
Here's my code:

```ts
import { type NextApiRequest, type NextApiResponse } from "next";
import { createNextApiHandler } from "@trpc/server/adapters/next";  <--- Here I import stuff from TRPC.
import { appRouter, createTRPCContext } from "@kdx/api"; // <--- Here I import stuff from my TRPC config.

// If you need to enable cors, you can do so like this:
const handler = (req: NextApiRequest, res: NextApiResponse) => {

  res.setHeader("Access-Control-Allow-Credentials", "*");
  // another common pattern
  res.setHeader("Access-Control-Allow-Origin", req.headers.origin || "*"); //NOTE: I also tried res.setHeader("Access-Control-Allow-Origin", "*"); and didnt work
  res.setHeader(
    "Access-Control-Allow-Methods",
    "GET,OPTIONS,PATCH,DELETE,POST,PUT",
  );
  res.setHeader(
    "Access-Control-Allow-Headers",
    "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version",
  );
  if (req.method === "OPTIONS") {
    res.status(200).end();
    return;
  }
  return createNextApiHandler({
    router: appRouter,
    createContext: createTRPCContext,
  })(req, res);
};
```
The above is working perfectly locally on my machine. (EG: server running on port 3000, and then another app I have running on port 3001 calling this endpoint on 3000).

**CHECKLIST**:
✔️ I have followed this link: https://vercel.com/guides/how-to-enable-cors and tried to set the following "headers" function to my next.config.js:
```
async headers() {
    return [
      {
        // matching all API routes
        source: "/api/:path*",
        headers: [
          { key: "Access-Control-Allow-Credentials", value: "true" },
          { key: "Access-Control-Allow-Origin", value: "*" },
          { key: "Access-Control-Allow-Methods", value: "GET,OPTIONS,PATCH,DELETE,POST,PUT" },
          { key: "Access-Control-Allow-Headers", value: "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version" },
        ]
      }
    ]
  }
  ```
  ✔️ I also tried to set the following to vercel.json:
```json
  {
  "headers": [
    {
      "source": "/api/(.*)",
      "headers": [
        { "key": "Access-Control-Allow-Credentials", "value": "true" },
        { "key": "Access-Control-Allow-Origin", "value": "*" },
        { "key": "Access-Control-Allow-Methods", "value": "GET,OPTIONS,PATCH,DELETE,POST,PUT" },
        { "key": "Access-Control-Allow-Headers", "value": "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version" }
      ]
    }
  ]
}
```

None of these solutions worked. I keep getting this error and I've seen multiple people online complain about this issue. I see that many have tried all of the steps suggested by vercel on https://vercel.com/guides/how-to-enable-cors, but many cannot solve it. I've even seen someone on twitter completely abandoning Vercel deployment and hosting their own server because of this.

**Comment by ngcxy** [comment]
I'm also facing the same problem. Just instead I'm deploying a React app connecting to a backend server. Configuring in vercel.json doesn't work at all. Is this Vercel's problem, not ours?

**Comment by dBianchii** [comment, marked as answer]
Hi everyone, I managed to fix it! Definitely not Vercels problem I think.
One issue that I was having was that I was trying to request https://(myurlhere) instead of https://www.(myurlhere)

I think that it errors when it realizes that there is a redirection. Also, pay attention to the error it throws on the browser window. It wasn't always the same to me, and it always had a very helpful CORS error message that I had to fix.
(E.G.): At one point, it was saying that I was incorrectly responding with the header Access-Control-Allow-Origin with the value '*', but that I could not send that when I had credentials set to 'include' in my api response. To fix it I had to send back in the Access-Control-Allow-Origin with the value of the website I was trying to request it from. (E.G): `req.headers.origin`.

So basically, to fix it I payed close attention to the specific CORS error the console was giving, and also by making sure the url I was requesting was the exact one, without any redirecting going on.

If anyone has any problems, please share the console error you are facing. Maybe I can be of help

**Reply by ngcxy** [reply]
Hi there. Congrats on solving the problem! I'm still working on mine.

In the beginning the error was blocked by CORS policy due to missing required header(even though I configured the vercel.json to add the header)

Then I tried adding "https://www" just like you do. It says: 

`POST (...my backend url with 'www'...) net::ERR_CERT_COMMON_NAME_INVALID`

I'm still learning web dev and this is my first time deploying a website. I really appreciate anything you can share

**Reply by dBianchii** [reply]
Could you share the code for your server route?

**Reply by ngcxy** [reply]
Of course: https://github.com/ngcxy/myBlog-backend

Here is the discussion link that includes further details: https://github.com/orgs/vercel/discussions/3279

**Reply by dBianchii** [reply]
Oh, I see. I don't know about django. Don't have any experience with it. My issue was inside a Next.js.

The way I fixed it was not by adding vercel.json but by adding the headers inside my api endpoint manually. Maybe fiddling with that can help you

**Reply by ngcxy** [reply]
I know it's a bit weird deploying Django on Vercel. It's an old project and I just want to move it from another platform to Vercel.

Now I finally work it out. I just realized it was a problem with the server. Adding the header didn't help at first. Then I find there's a 'CORS_ALLOWED_ORIGINS' variable in settings, and changing it did help.

Thank you so much for the help.

**Reply by irohitb** [reply]
@ngcxy 

Where in setting is `CORS_ALLOWED_ORIGINS`?

**Reply by yashcrest** [reply]
Hey @dBianchii , sorry I might be late to the party. But I am running into the same problem that you faced previously. 

I have removed the vercel.json headers config file and am handling the headers inside my api file itself as well. Could you care to maybe have a look and point out where I am messing it up. 
Here are the links to my repo: 
Frontend: https://github.com/yashcrest/react-e-commerce
Backend: https://github.com/yashcrest/react-e-commerce-backend

<img width="756" alt="Screenshot 2023-11-09 at 6 48 16 pm" src="https://github.com/vercel/community/assets/79971012/bc33f18e-df7b-4cb5-8592-de7da4e472cf">

**Reply by thearpitsarap** [reply]
This worked properly. The CORS error has now fixed, but I am getting net::ERR_CERT_COMMON_NAME_INVALID when added www before https://(myurl). Any solution for this?

**Reply by lelouchB** [reply]
I was facing the exact issue for my https://finalspaceapi.com 
It turns out I was redirecting all requests to www while using the API without the www.
Removed the redirects and it started working

**Reply by therealeddy** [reply]
You are amazing bro, thanks!

**Comment by kyle-ssg** [comment]
After going through every GH issue and trying all of the above, I figured out in the end that I had to add a trailing / to the endpoint I was attempting to hit.

I figured this out by reading the cors issue in the browser more carefully.  'preflight is invalid (redirect)' or 'redirect is not allowed for a preflight request'. This was because Next.js was redirecting to the same URL with an appended /.

**Comment by whalyf** [comment]
First CORS-ORIGIN error, then PREFLIGHT error, i've solved that adding rewrites and many headers allowed to my ``vercel.json`` file.
My api is just a simple serveless function in golang. You guys should try that, worked for me.

```
 {
  "version": 2,
  "build": {
    "env": {
      "GO_BUILD_FLAGS": "-ldflags '-s -w'"
    }
  },
  "rewrites": [
    {
      "source": "/api/(.*)",
      "destination": "/api/main.go"
    }
  ],
  "headers": [
    {
      "source": "/api/(.*)",
      "headers": [
        { "key": "Access-Control-Allow-Credentials", "value": "true" },
        { "key": "Access-Control-Allow-Origin", "value": "*" },
        { "key": "Access-Control-Allow-Methods", "value": "GET,OPTIONS,PATCH,DELETE,POST,PUT" },
        { "key": "Access-Control-Allow-Headers", "value": "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version" }
      ]
    }
  ]
}
```

**Comment by najwer23** [comment]
My React APP host: http://localhost:3000
My serverless APP host: http://localhost:5000

And it works with Vercel CLI 31.1.1 (npm i -g vercel@31.1.1) without extra configuration on Vercel

But it doesnt work with Vercel CLI 34.1.8

why? What happend after 31.1.1?

**Comment by Khaleeq01** [comment]
I deoloyed my frontend and backend on vercel. But i face error in frontend can someone help me?
food-delivery-website-gamma.vercel.app/:1 Access to XMLHttpRequest at 'https://food-delivery-website-backend.vercel.app/api/food/list' from origin 'https://food-delivery-website-gamma.vercel.app' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
Failed to load resource: net::ERR_FAILED
Uncaught (in promise) Error: A listener indicated an asynchronous response by returning true, but the message channel closed before a response was received
